### PR TITLE
Add data requirements op and measure bundle assembly

### PR DESCRIPTION
**Pull request by mgramigna**
# Summary

Updates to the functionality introduced in #10, but on a clean branch with added support for $data-requirements

## New behavior

Adds the `$data-requirements` operation

## Code changes

* Update code to assemble measure bundle from measure ID
* Add custom op for data requirements that calls `fqm-execution`'s `$data-requirements` function
* Update dependencies with lodash and fqm-e

# Testing guidance

POST a measure bundle from the `connectathon` repo into your database using regular transaction bundle upload. Note the measure ID of the created Measure resource (e.g. `measure-EXM130-7.3.000`).

Send GET request to `http://localhost:3000/4_0_0/Measure/<your-measure-id>/$data-requirements`

Output should be a FHIR Library `module-definition` with all the data requirements

### CODE REVIEW

File: src/util/measureBundle.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+  // This filter checks for the 'Library' keyword on all related artifacts
+  // TODO: This filter can probably be improved, but will work in our cases for now
+  const depLibUrls = lib.relatedArtifact
+    .filter(ra => ra.type === 'depends-on' && ra.resource.includes('Library'))
+    .map(ra => ra.resource);
+
+  // Obtain all libraries referenced in the related artifact, and recurse on their dependencies
+  const libraryGets = depLibUrls.map(async url => {
+    const [urlPart, versionPart] = url.split('|');
```
Comment (hossenlopp) [inline src/util/measureBundle.js:96]: This doesn't handle the situation where no version is specified on the URL.

File: src/util/measureBundle.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+      ]
+    });
+  }
+
+  const [mainLibraryRef] = measure.library;
+
+  // Library references could be canonical or resourceType/id
+  let mainLib;
+  if (isValidLibraryUrl(mainLibraryRef)) {
+    mainLib = await findOneResourceWithQuery({ url: mainLibraryRef }, 'Library');
```
Comment (hossenlopp) [inline src/util/measureBundle.js:48]: This doesn't handle the situation where the canonical url includes the version.

File: src/util/measureBundle.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  const libraryGets = depLibUrls.map(async url => {
+    const [urlPart, versionPart] = url.split('|');
+
+    const lib = await findOneResourceWithQuery({ url: urlPart, version: versionPart }, 'Library');
+    return getAllDependentLibraries(lib);
+  });
+
+  const allDeps = await Promise.all(libraryGets);
+
+  results.push(...allDeps);
```
Comment (hossenlopp) [inline src/util/measureBundle.js:104]: There could be the possibility that this grabs duplicate copies of libraries if multiple supporting libraries use the same libraries.

**Review by hossenlopp — CHANGES_REQUESTED** [review]
`$data-requirements` call works but there could be some issues with finding supporting libraries. The bundles for each measure do not have the ValueSets which will be needed for subsequent execution/gaps work.

File: src/util/measureBundle.js (hunk 1 above)
Comment (mgramigna) [reply src/util/measureBundle.js]: Good catch, added a function to support this

File: src/util/measureBundle.js (hunk 2 above)
Comment (mgramigna) [reply src/util/measureBundle.js]: Good catch, added a function to support this

File: src/util/measureBundle.js (hunk 3 above)
Comment (mgramigna) [reply src/util/measureBundle.js:142]: Yeah.. the duplicate logic is not ideal. In the interest of time, we decided to leave it. I noted a TODO in there to improve the recursion to filter out already visited resources.

**Review by hossenlopp — APPROVED** [review]
Looks good to me. Now do execution and care gaps. 🤣

**Pull request merged**